File: LeetCodeDaily/26_shortestBeutifullSubstringExactltK1s.py
def solve(s,k):
    n = len(s)

    bestLen = float('inf')
    ans = ""
    ones = 0

    l = 0

    for r in range(n):
        if s[r] == '1':
            ones += 1

        while ones >= k:

            if ones == k:
                currLen = r-l+1
                curr = s[l:r+1]
                if currLen < bestLen:
                    bestLen = currLen
                    ans = curr
                elif currLen == bestLen:
                    ans = min(ans,curr)
            if s[l] == '1':
                ones -= 1
            l += 1
    return ans


def main():
    s = input()
    k = int(input())

    print(solve(s,k))

main()
    

File: LeetCodeDaily/26_shortestBeutifullSubstringExactltK1s.cpp
#include<bits/stdc++.h>
using namespace std;
// 
string shortestBeautifulSubstring(string s, int k) {
    int n = s.length();
    
    int bestLen = INT_MAX;
    string ans = "";

    int ones = 0;

    int l = 0;

    for(int r=0;r<n;r++){
        if(s[r]=='1') ones++;

        while (ones >= k)
        {
            if(ones == k){
                int len = r-l+1;
                string curr = s.substr(l,len);
                
                if(len < bestLen){
                    bestLen = len;
                    ans = curr;
                }else if(len == bestLen){
                    ans = min(curr,curr);
                }
            }

            if(s[l]=='1'){
                ones--;
            }
            l++;
        }
        
    }
    return ans;
}

int main(){
    string s;
    cin >> s;
    int k;
    cin >> k;
    cout << shortestBeautifulSubstring(s,k);
    return 0;
}